pkg/sources/filesystem/filesystem.go

```diff
@@ -93,18 +113,70 @@ func (s *Source) Chunks(ctx context.Context, chunksChan chan *sources.Chunk, _ .
 		}
 		s.SetProgressComplete(i, len(s.paths), fmt.Sprintf("Path: %s", path), "")
 
+		// Initialize per-path tracking - critically important for memory management.
+		// scanRootPaths is reset for each top-level path to track depth-1 symlinks.
+		s.scanRootPaths = make(map[string]struct{})
+
+		// Create LRU cache only if following symlinks to avoid unnecessary memory allocation.
+		// The cache is recreated for each scan path to prevent memory accumulation across
+		// multiple scans. This ensures O(paths_per_scan) memory instead of O(total_paths).
+		if s.followSymlinks {
+			// Maximum of 10k paths limits memory to ~1MB even for very large directory trees.
+			// If a directory has >10k symlinks, oldest entries are evicted (LRU behavior).
+			const maxCacheSize = 10000
+			cache, err := lru.NewCache[struct{}]("filesystem_visited", lru.WithCapacity[struct{}](maxCacheSize))
+			if err != nil {
+				logger.Error(err, "failed to create LRU cache for symlink tracking")
+				continue
+			}
+			s.visitedPaths = cache
+		}
+
 		cleanPath := filepath.Clean(path)
-		fileInfo, err := os.Lstat(cleanPath)
+
+		// Store the scan root path for depth tracking
+		s.scanRootPaths[cleanPath] = struct{}{}
+
+		var fileInfo fs.FileInfo
+		var err error
+		if s.followSymlinks {
+			fileInfo, err = os.Stat(cleanPath)
+		} else {
+			fileInfo, err = os.Lstat(cleanPath)
+		}
 		if err != nil {
 			logger.Error(err, "unable to get file info")
 			continue
 		}
 
-		if fileInfo.Mode()&os.ModeSymlink != 0 {
+		if !s.followSymlinks && fileInfo.Mode()&os.ModeSymlink != 0 {
 			logger.Info("skipping, not a regular file", "path", cleanPath)
 			continue
 		}
 
+		// If followSymlinks is enabled and this is a symlink, check for loops
+		if s.followSymlinks && fileInfo.Mode()&os.ModeSymlink != 0 {
```

Comment:
I believe this is *always* false since `fileInfo` comes from `Stat` when `followSymlinks` is true. Is there an example or test case that exercises this scenario?